File: sdks/typescript/chio-ts/src/invariants/receipt.ts
import { canonicalizeJson, canonicalizeJsonString } from "./json.ts";
import { publicKeyHexMatches, sha256Hex, verifyChioSignature } from "./crypto.ts";
import { ChioInvariantError, parseJsonText } from "./errors.ts";

function safeVerifyReceiptSignature(
  signedBytes: string,
  signature: string,
  publicKey: string,
): boolean {
  try {
    return verifyChioSignature(signedBytes, signature, publicKey);
  } catch (error) {
    if (error instanceof ChioInvariantError) {
      // Fail-closed: receipts signed with unsupported algorithms or malformed
      // keys/signatures verify as false rather than propagating an exception.
      return false;
    }
    throw error;
  }
}

export type ReceiptDecisionKind = "allow" | "deny" | "cancelled" | "incomplete" | "none";

export interface ToolCallAction {
  parameters: unknown;
  parameter_hash: string;
}

export interface ChioReceipt {
  id: string;
  timestamp: number;
  capability_id: string;
  tool_server: string;
  tool_name: string;
  action: ToolCallAction;
  decision?: {
    verdict: ReceiptDecisionKind;
    reason?: string;
    guard?: string;
  };
  receipt_kind: string;
  boundary_class: string;
  observation_outcome?: string;
  tool_origin: string;
  redaction_mode: string;
  actor_chain?: Array<{
    actor_id: string;
    actor_kind?: string;
  }>;
  content_hash: string;
  policy_hash: string;
  evidence?: Array<{
    guard_name: string;
    verdict: boolean;
    details?: string;
  }>;
  metadata?: unknown;
  trust_level: "mediated" | "verified" | "advisory" | string;
  tenant_id?: string;
  kernel_key: string;
  algorithm?: unknown;
  signature: string;
}

export interface ReceiptVerification {
  signature_valid: boolean;
  parameter_hash_valid: boolean;
  receipt_id_valid: boolean;
  decision: ReceiptDecisionKind;
  receipt_kind: string;
  boundary_class: string;
  trust_level: string;
  result: string;
  authorized: boolean;
  signer_key_hex: string;
  signer_trusted: boolean;
  ok: boolean;
}

export function parseReceiptJson(input: string): ChioReceipt {
  return parseJsonText(input);
}

export function receiptBody(receipt: ChioReceipt): Omit<ChioReceipt, "algorithm" | "signature"> {
  const { signature: _signature, algorithm: _algorithm, ...body } = receipt;
  return body;
}

export function receiptBodyCanonicalJson(receipt: ChioReceipt): string {
  return canonicalizeJson(receiptBody(receipt));
}

function receiptIdInput(receipt: ChioReceipt): Record<string, unknown> {
  const input: Record<string, unknown> = {
    action: receipt.action,
    capability_id: receipt.capability_id,
    content_hash: receipt.content_hash,
    receipt_kind: receipt.receipt_kind,
    boundary_class: receipt.boundary_class,
    tool_origin: receipt.tool_origin,
    redaction_mode: receipt.redaction_mode,
    kernel_key: receipt.kernel_key,
    policy_hash: receipt.policy_hash,
    timestamp: receipt.timestamp,
    tool_name: receipt.tool_name,
    tool_server: receipt.tool_server,
  };
  if (receipt.evidence !== undefined && receipt.evidence.length > 0) {
    input.evidence = receipt.evidence;
  }
  if (receipt.decision !== undefined) {
    input.decision = receipt.decision;
  }
  if (receipt.observation_outcome !== undefined) {
    input.observation_outcome = receipt.observation_outcome;
  }
  if (receipt.actor_chain !== undefined && receipt.actor_chain.length > 0) {
    input.actor_chain = receipt.actor_chain;
  }
  if (receipt.metadata !== undefined && receipt.metadata !== null) {
    input.metadata = receipt.metadata;
  }
  input.trust_level = receipt.trust_level;
  if (receipt.tenant_id !== undefined) {
    input.tenant_id = receipt.tenant_id;
  }
  return input;
}

function contentAddressedReceiptId(receipt: ChioReceipt): string {
  return sha256Hex(canonicalizeJson(receiptIdInput(receipt)));
}

export function receiptSigningBodyCanonicalJson(receipt: ChioReceipt): string {
  return canonicalizeJson({
    id: receipt.id,
    body: receiptIdInput(receipt),
  });
}

function receiptSemantics(receipt: ChioReceipt): { receipt_kind: string; boundary_class: string } {
  return { receipt_kind: receipt.receipt_kind, boundary_class: receipt.boundary_class };
}

function resultLabel(receiptKind: string, boundaryClass: string, decision: ReceiptDecisionKind): string {
  if (receiptKind === "mediated_decision" && boundaryClass === "prevent" && decision === "allow") {
    return "Authorized";
  }
  if (receiptKind === "trace_observation") {
    return "Observed";
  }
  if (receiptKind === "advisory_evaluation") {
    return "Advisory";
  }
  switch (decision) {
    case "allow":
      return "Allowed";
    case "deny":
      return "Denied";
    case "cancelled":
      return "Cancelled";
    case "incomplete":
      return "Incomplete";
    case "none":
      return "Invalid";
  }
}

function validObservationOutcome(value: string | undefined): boolean {
  return value === "observed" || value === "evaluated" || value === "dropped";
}

function semanticallySignable(receipt: ChioReceipt, decision: ReceiptDecisionKind): boolean {
  if (receipt.receipt_kind === "mediated_decision") {
    return (
      receipt.boundary_class === "prevent" &&
      receipt.trust_level === "mediated" &&
      decision !== "none" &&
      receipt.observation_outcome === undefined
    );
  }
  if (receipt.receipt_kind === "trace_observation") {
    return (
      receipt.boundary_class === "detect_only" &&
      receipt.trust_level === "verified" &&
      decision === "none" &&
      validObservationOutcome(receipt.observation_outcome)
    );
  }
  if (receipt.receipt_kind === "advisory_evaluation") {
    return (
      receipt.boundary_class === "advisory_only" &&
      receipt.trust_level === "advisory" &&
      decision === "none" &&
      validObservationOutcome(receipt.observation_outcome)
    );
  }
  return false;
}

export function verifyReceipt(receipt: ChioReceipt, trustedSigners: string[] = []): ReceiptVerification {
  const signingBodyCanonicalJson = receiptSigningBodyCanonicalJson(receipt);
  const parameterCanonicalJson = canonicalizeJson(receipt.action.parameters);
  const decision = receipt.decision?.verdict ?? "none";
  const semantics = receiptSemantics(receipt);
  const semanticAuthorized =
    semantics.receipt_kind === "mediated_decision" && semantics.boundary_class === "prevent" && decision === "allow";
  const signerTrusted =
    trustedSigners.length > 0 && trustedSigners.some((signer) => publicKeyHexMatches(signer, receipt.kernel_key));
  const receiptIdValid = receipt.id === contentAddressedReceiptId(receipt);
  const signatureValid =
    receiptIdValid &&
    semanticallySignable(receipt, decision) &&
    safeVerifyReceiptSignature(signingBodyCanonicalJson, receipt.signature, receipt.kernel_key);
  const parameterHashValid = receipt.action.parameter_hash === sha256Hex(parameterCanonicalJson);
  const authorized = semanticAuthorized && signatureValid && parameterHashValid && receiptIdValid && signerTrusted;

  return {
    signature_valid: signatureValid,
    parameter_hash_valid: parameterHashValid,
    receipt_id_valid: receiptIdValid,
    decision,
    receipt_kind: semantics.receipt_kind,
    boundary_class: semantics.boundary_class,
    trust_level: receipt.trust_level,
    result: resultLabel(semantics.receipt_kind, semantics.boundary_class, decision),
    authorized,
    signer_key_hex: receipt.kernel_key,
    signer_trusted: signerTrusted,
    ok: signatureValid && parameterHashValid && receiptIdValid && signerTrusted,
  };
}

export function verifyReceiptWithTrustedSigners(
  receipt: ChioReceipt,
  trustedSigners: string[],
): ReceiptVerification {
  return verifyReceipt(receipt, trustedSigners);
}

export function verifyReceiptJson(input: string): ReceiptVerification {
  return verifyReceipt(parseReceiptJson(input));
}

export { canonicalizeJsonString };

File: sdks/python/chio-py/src/chio/invariants/receipt.py
from __future__ import annotations

from typing import Any

from ..errors import ChioInvariantError, parse_json_text
from .hashing import sha256_hex_utf8
from .json import canonicalize_json
from .signing import (
    public_key_hex_matches,
    verify_chio_signature,
)


def _safe_verify_receipt_signature(
    signed_text: str,
    signature: str,
    public_key: str,
) -> bool:
    """Dispatch through ``verify_chio_signature``, swallowing algorithm errors.

    Receipts signed with algorithms this SDK build does not yet support fail
    closed by reporting ``signature_valid: False`` rather than propagating an
    exception, mirroring the TS behaviour.
    """

    try:
        return verify_chio_signature(signed_text, signature, public_key)
    except ChioInvariantError:
        return False


def parse_receipt_json(input_text: str) -> dict[str, Any]:
    return parse_json_text(input_text)


def _receipt_body(receipt: dict[str, Any]) -> dict[str, Any]:
    return {
        key: value
        for key, value in receipt.items()
        if key not in {"algorithm", "signature"}
    }


def receipt_body_canonical_json(receipt: dict[str, Any]) -> str:
    return canonicalize_json(_receipt_body(receipt))


def _receipt_id_input(receipt: dict[str, Any]) -> dict[str, Any]:
    value: dict[str, Any] = {
        "timestamp": receipt["timestamp"],
        "capability_id": receipt["capability_id"],
        "tool_server": receipt["tool_server"],
        "tool_name": receipt["tool_name"],
        "action": receipt["action"],
        "receipt_kind": receipt["receipt_kind"],
        "boundary_class": receipt["boundary_class"],
        "tool_origin": receipt["tool_origin"],
        "redaction_mode": receipt["redaction_mode"],
        "content_hash": receipt["content_hash"],
        "policy_hash": receipt["policy_hash"],
        "kernel_key": receipt["kernel_key"],
    }
    if receipt.get("decision") is not None:
        value["decision"] = receipt["decision"]
    if receipt.get("observation_outcome") is not None:
        value["observation_outcome"] = receipt["observation_outcome"]
    if receipt.get("actor_chain"):
        value["actor_chain"] = receipt["actor_chain"]
    if receipt.get("evidence"):
        value["evidence"] = receipt["evidence"]
    if receipt.get("metadata") is not None:
        value["metadata"] = receipt["metadata"]
    value["trust_level"] = receipt["trust_level"]
    if receipt.get("tenant_id") is not None:
        value["tenant_id"] = receipt["tenant_id"]
    return value


def _content_addressed_receipt_id(receipt: dict[str, Any]) -> str:
    return sha256_hex_utf8(canonicalize_json(_receipt_id_input(receipt)))


def receipt_signing_body_canonical_json(receipt: dict[str, Any]) -> str:
    return canonicalize_json({
        "id": receipt["id"],
        "body": _receipt_id_input(receipt),
    })


def _receipt_semantics(receipt: dict[str, Any]) -> tuple[str, str]:
    return receipt["receipt_kind"], receipt["boundary_class"]


def _result_label(receipt_kind: str, boundary_class: str, decision: str) -> str:
    if receipt_kind == "mediated_decision" and boundary_class == "prevent" and decision == "allow":
        return "Authorized"
    if receipt_kind == "trace_observation":
        return "Observed"
    if receipt_kind == "advisory_evaluation":
        return "Advisory"
    return {
        "allow": "Allowed",
        "deny": "Denied",
        "cancelled": "Cancelled",
        "incomplete": "Incomplete",
        "none": "Invalid",
    }[decision]


def _valid_observation_outcome(value: Any) -> bool:
    return value in {"observed", "evaluated", "dropped"}


def _semantically_signable(receipt: dict[str, Any], decision: str) -> bool:
    receipt_kind, boundary_class = _receipt_semantics(receipt)
    trust_level = receipt.get("trust_level")
    observation_outcome = receipt.get("observation_outcome")
    if receipt_kind == "mediated_decision":
        return (
            boundary_class == "prevent"
            and trust_level == "mediated"
            and decision != "none"
            and observation_outcome is None
        )
    if receipt_kind == "trace_observation":
        return (
            boundary_class == "detect_only"
            and trust_level == "verified"
            and decision == "none"
            and _valid_observation_outcome(observation_outcome)
        )
    if receipt_kind == "advisory_evaluation":
        return (
            boundary_class == "advisory_only"
            and trust_level == "advisory"
            and decision == "none"
            and _valid_observation_outcome(observation_outcome)
        )
    return False


def verify_receipt(
    receipt: dict[str, Any],
    trusted_signers: list[str] | None = None,
) -> dict[str, Any]:
    decision_value = receipt.get("decision")
    decision = (
        decision_value.get("verdict", "none")
        if isinstance(decision_value, dict)
        else "none"
    )
    receipt_kind, boundary_class = _receipt_semantics(receipt)
    semantic_authorized = receipt_kind == "mediated_decision" and boundary_class == "prevent" and decision == "allow"
    receipt_id_valid = receipt["id"] == _content_addressed_receipt_id(receipt)
    signature_valid = (
        receipt_id_valid
        and _semantically_signable(receipt, decision)
        and _safe_verify_receipt_signature(
            receipt_signing_body_canonical_json(receipt),
            receipt["signature"],
            receipt["kernel_key"],
        )
    )
    parameter_hash_valid = receipt["action"]["parameter_hash"] == sha256_hex_utf8(
        canonicalize_json(receipt["action"]["parameters"])
    )
    signer_trusted = bool(trusted_signers) and any(
        public_key_hex_matches(signer, receipt["kernel_key"])
        for signer in trusted_signers or []
    )
    authorized = (
        semantic_authorized
        and signature_valid
        and parameter_hash_valid
        and receipt_id_valid
        and signer_trusted
    )
    return {
        "signature_valid": signature_valid,
        "parameter_hash_valid": parameter_hash_valid,
        "receipt_id_valid": receipt_id_valid,
        "decision": decision,
        "receipt_kind": receipt_kind,
        "boundary_class": boundary_class,
        "trust_level": receipt["trust_level"],
        "result": _result_label(receipt_kind, boundary_class, decision),
        "authorized": authorized,
        "signer_key_hex": receipt["kernel_key"],
        "signer_trusted": signer_trusted,
        "ok": signature_valid and parameter_hash_valid and receipt_id_valid and signer_trusted,
    }


def verify_receipt_with_trusted_signers(
    receipt: dict[str, Any],
    trusted_signers: list[str],
) -> dict[str, Any]:
    return verify_receipt(receipt, trusted_signers)


def verify_receipt_json(input_text: str) -> dict[str, Any]:
    return verify_receipt(parse_receipt_json(input_text))
